### PR TITLE
feat!(module): Adds developer fee sharing mechanism

### PR DESCRIPTION
**Pull request by Reecepbcups**
WIP

Adds cosmwasm fee share split after fees are burned. I ported over from my [Juno v12 PR](https://github.com/CosmosContracts/juno/pull/385). 

Needs:
- Upgrade Handler (has the v2 upgrade handler already run? if so I need to move to a new v2_0_4 upgrade)
- e2e test

### CODE REVIEW

**Comment by faddat** [comment]
image

look before you leap

**Comment by faddat** [comment]
read before you leap

https://docs.google.com/document/d/1AIOOrHiNAFQKwzaeDNyJJJx8SwXNlaSUcZbasnezXS4/edit?usp=sharing

**Comment by faddat** [comment]
I'm aware that you could, but -- be informed first

https://docs.google.com/document/d/1lDGQiUBbkwBPyRLFA3dnckBBHvYSL51JoqRlgPcCftA/edit?usp=sharing

File: x/feeshare/ante/ante_test.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+	suite.Suite
+}
+
+func TestAnteSuite(t *testing.T) {
+	suite.Run(t, new(AnteTestSuite))
+}
+
+func (suite *AnteTestSuite) TestFeeLogic() {
+	// We expect all to pass
+	feeCoins := sdk.NewCoins(sdk.NewCoin("ujuno", sdk.NewInt(500)), sdk.NewCoin("utoken", sdk.NewInt(250)))
```
Comment (ZaradarBH) [inline x/feeshare/ante/ante_test.go:22]: Might want to change these to uluna :)

File: x/feeshare/genesis_test.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+			suite.genesis,
+			false,
+		},
+		{
+			"custom genesis - feeshare disabled",
+			types.GenesisState{
+				Params: types.Params{
+					EnableFeeShare:  false,
+					DeveloperShares: types.DefaultDeveloperShares,
+					AllowedDenoms:   []string{"ujuno"},
```
Comment (ZaradarBH) [inline x/feeshare/genesis_test.go:57]: Change to uluna :)

File: x/feeshare/keeper/grpc_query.go [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+	}
+
+	ctx := sdk.UnwrapSDKContext(c)
+
+	// check if the contract is a non-zero hex address
+	contract, err := sdk.AccAddressFromBech32(req.ContractAddress)
+	if err != nil {
+		return nil, status.Errorf(
+			codes.InvalidArgument,
+			"invalid format for contract %s, should be bech32 ('juno...')", req.ContractAddress,
```
Comment (ZaradarBH) [inline x/feeshare/keeper/grpc_query.go:76]: Change to terra :)

File: x/feeshare/keeper/grpc_query.go [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+		return nil, status.Error(codes.InvalidArgument, "empty request")
+	}
+
+	ctx := sdk.UnwrapSDKContext(c)
+
+	deployer, err := sdk.AccAddressFromBech32(req.DeployerAddress)
+	if err != nil {
+		return nil, status.Errorf(
+			codes.InvalidArgument,
+			"invalid format for deployer %s, should be bech32 ('juno...')", req.DeployerAddress,
```
Comment (ZaradarBH) [inline x/feeshare/keeper/grpc_query.go:118]: Change to terra :)

File: x/feeshare/keeper/grpc_query.go [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+		return nil, status.Error(codes.InvalidArgument, "empty request")
+	}
+
+	ctx := sdk.UnwrapSDKContext(c)
+
+	deployer, err := sdk.AccAddressFromBech32(req.WithdrawerAddress)
+	if err != nil {
+		return nil, status.Errorf(
+			codes.InvalidArgument,
+			"invalid format for withdraw addr %s, should be bech32 ('juno...')", req.WithdrawerAddress,
```
Comment (ZaradarBH) [inline x/feeshare/keeper/grpc_query.go:157]: Change to terra :)

File: x/feeshare/spec/01_concepts.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
+<!--
+order: 1
+-->
+
+# Concepts
+
+## FeeShare
+
+The FeeShare module is a revenue-per-transaction model, which allows developers to get paid for deploying their decentralized applications (dApps) on Juno. This helps developers to generate revenue every time a user interacts with their contracts on the chain. This registration is permissionless to sign up for and begin earning fees from. By default, 50% of all transaction fees for Execute Messages are shared. This can be changed by governance and implemented by the `x/feeshare` module.
```
Comment (ZaradarBH) [inline x/feeshare/spec/01_concepts.md:9]: Change Juno to Terra Classic :)

File: x/feeshare/spec/01_concepts.md [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+## Registration
+
+Developers register their contract applications to gain their cut of fees per execution. Any contract can be registered by a developer by submitting a signed transaction. After the transaction is executed successfully, the developer will start receiving a portion of the transaction fees paid when a user interacts with the registered contract. The developer can have the funds sent to their wallet, a DAO, or any other wallet address on the Juno Network.
```
Comment (ZaradarBH) [inline x/feeshare/spec/01_concepts.md:13]: Change Juno to Terra Classic :)

File: x/feeshare/spec/README.md [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  title: "feeshare"
+-->
+
+# `feeshare`
+
+## Abstract
+
+This document specifies the internal `x/feeshare` module of Juno Network.
+
+The `x/feeshare` module enables the Juno to support splitting transaction fees between the community and smart contract deployer. This aims to increase the adoption of Juno by offering a new way of income for CosmWasm smart contract developers. Developers can register their smart contracts and every time someone interacts with a registered smart contract, the contract deployer or their assigned withdrawal account receives a part of the transaction fees.
```
Comment (ZaradarBH) [inline x/feeshare/spec/README.md:14]: Change Juno to Terra Classic :)

File: x/feeshare/spec/README.md [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+title: "FeeShare Overview"
+parent:
+  title: "feeshare"
+-->
+
+# `feeshare`
+
+## Abstract
+
+This document specifies the internal `x/feeshare` module of Juno Network.
```
Comment (ZaradarBH) [inline x/feeshare/spec/README.md:12]: Change Juno to Terra Classic :)

File: app/app.go [hunk 10]
```diff
@@ -137,6 +137,9 @@ import (
 	wasmkeeper "github.com/terra-money/core/x/wasm/keeper"
 	wasmtypes "github.com/terra-money/core/x/wasm/types"
 
+	feesharekeeper "github.com/terra-money/core/x/feeshare/keeper"
```
Comment (ZaradarBH) [inline app/app.go:140]: Change "terra-money" namespace to "classic-terra"

**Review by ZaradarBH — COMMENTED** [review]
Good work. I have commented on a few ad-hoc cosmetics. Also it looks like the protogogo tooling broke (again) :)

File: app/app.go (hunk 10 above)
Comment (Reecepbcups) [reply app/app.go:140]: I don't believe this is possible with the go module still being `module github.com/terra-money/core`. Unless that is what you want me to do here for all other lines in the codebase as well

Comment (ZaradarBH) [reply app/app.go:140]: Merge this branch into yours and you should be good to go @ https://github.com/classic-terra/core/tree/correct-module-path :)

Comment (Reecepbcups) [reply app/app.go]: resolved in ff31c7bc3d3143808915f434139a5e4330caf1ab

**Comment by ZaradarBH** [comment]
Your a machine buddy! :) Looks good now. I will discuss it with the team today and get back to you with a estimate on when we will merge in your suggestion.

**Comment by Reecepbcups** [comment]
Hanks @ZaradarBH. I still do need to test this locally, will mark the PR ready for review when It’s good

Also going to add a native tokenfactory for v2.0.4 as well. 

Just slipping in PRs when I have time

**Comment by inon-man** [comment]
Please target the base branch to release/1.x. We are going to abandon the current main.

**Comment by fragwuerdig** [comment]
@Reecepbcups: I appreciate your effort and am REALLY happy about you helping us out here. Would you be so kind to rebase your changes on the main branch? Your branch and our main seem kinda diverged?

**Comment by Reecepbcups** [comment]
@fragwuerdig I am so confused. Current main uses SDK v0.44.8, so things are rolling back from v45.11? I merged in the release/v1.x branch and that is what broke this PR. A `git revert -m 1 HEAD` fixes that.

Then a merge of main into this one. But going backwards here seems wrong. What am I missing? is this intended?

**Comment by fragwuerdig** [comment]
@Reecepbcups @inon-man: We abandoned the old main, because it was unstable. During the establishment of this team we did not have a clear pull request policy. @inon-man saved us from going wild ;).

SDK 45 was never part of any release, so I figure we are not really reverting from anything beefy here. On top of that: The release series v1.x will be abandoned too after v1.1.0 upgrade, which is already up for proposal.

The juicy commits about fee sharing are the ones starting with commit hash `a6274f2`, is that right? So it would be cool to make an interactive rebase on main and drop everything except these? Then force push?

If force push is no option, then second option would be to recreate based on main and cherry pick the juicy commits?

@inon-man, what is your suggestion?

So sorry for the confusion and the overhead :( It's really our bad...

**Comment by Reecepbcups** [comment]
It actually starts at commit [a8709a7](https://github.com/classic-terra/core/pull/75/commits/a8709a76dbd31840a8378e4bc139c78782fc8224) . This is where I bring in all the module code initially

**Comment by fragwuerdig** [comment]
> It actually starts at commit [a8709a7](https://github.com/classic-terra/core/pull/75/commits/a8709a76dbd31840a8378e4bc139c78782fc8224) . This is where I bring in all the module code initially

Aaaaah. Yeah, I see. Sorry, it was hidden behind a series of expansion widgets :)

**Comment by Reecepbcups** [comment]
@inon-man yes! I will get to this later today / middle of the day tomorrow :) thanks

**Pull request closed without merging**